### PR TITLE
merge 2.13.perfopt branch to master branch

### PR DESCRIPTION
**Pull request by immarvin**
For task https://github.com/xcat2/xcat-core/issues/2907, merge https://github.com/xcat2/xcat-core/tree/2.13.perfopt branch to master branch:

1. push all the ``makedhcp`` enhancement code into master branch
2. push the xcat perf code into master branch

### CODE REVIEW

**Comment by whowutwut** [comment]
Is this kind of a last minute merge to the release branch kind of risky the day before starting release branch testing?  Should we have some testing data included with the pull request to ensure that we did not break anything.

**Comment by immarvin** [comment]
hi @whowutwut , I ran autotest of 2.13.perfopt branch against the following <os+arch>, the result:

````
[rhels6.9+x86_64]

Start time: 2017-05-12 20:57:52

End time  : 2017-05-13 04:42:32

All time consumption: 7 hours 44 minutes 40 seconds

Deploy test environment time consumption: 0 hours 8 minutes 36 seconds

xcat core : 2.13.3-snap201705090411

xcat dep  : http://xcat.org/files/xcat/xcat-dep/2.x_Linux/xcat-dep-201702211334.tar.bz2

commit num: 25bbef0

TotalCase 232 Pass 207 Failed 25 NoRun 0

Failed cases: reg_linux_diskfull_installation_flat, go_xcat_local_repo_case7, go_xcat_noinput, go_xcat_with_x, go_xcat_with_xcat-version-1, go_xcat_online_repo_case6, xdsh_regular_command, xdsh_Q_command, xdsh_c_cn, xdsh_e_filename, xdsh_E, xdsh_t, xdsh_T, xdsh_o, psh_node_cmd_linux, psh_l, psh_i, prsync_dir_node, prsync_file_node, reg_linux_diskless_installation_flat, packimage_m_cpio_c_gzip, packimage_m_cpio_c_xz, packimage_m_tar_c_gzip, packimage_m_tar_c_xz, reg_linux_statelite_installation_flat,

````

````
 
======================================
  Test Result for Project sles12.2-ppc64le-makedhcp-5
======================================

[sles12.2+ppc64le]

Start time: 2017-05-12 00:46:28

End time  : 2017-05-12 09:31:35

All time consumption: 8 hours 45 minutes 7 seconds

Deploy test environment time consumption: 0 hours 25 minutes 38 seconds

xcat core : 2.13.3-snap201705090411

xcat dep  : http://xcat.org/files/xcat/xcat-dep/2.x_Linux/xcat-dep-201702211334.tar.bz2

commit num: 25bbef0

TotalCase 240 Pass 222 Failed 18 NoRun 0

Failed cases: reg_linux_diskless_installation_flat, go_xcat_noinput, go_xcat_with_x, go_xcat_with_xcat-version-1, go_xcat_online_repo_case6, packimage_m_cpio_c_gzip, packimage_m_cpio_c_pigz, packimage_m_cpio_c_xz, packimage_m_tar_c_pigz, packimage_m_tar_c_gzip, packimage_m_tar_c_xz, makedns_d_node, makedns_n, makedns, xdsh_t, sles_migration1, sles_migration2, reg_linux_diskless_installation_hierarchy,


````

````
----- Original message -----

 
======================================
  Test Result for Project sles11.4-x86_64-makedhcp-1
======================================

[sles11.4+x86_64]

Start time: 2017-05-12 12:57:45

End time  : 2017-05-12 20:56:28

All time consumption: 7 hours 58 minutes 43 seconds

Deploy test environment time consumption: 0 hours 10 minutes 9 seconds

xcat core : 2.13.3-snap201705090411

xcat dep  : http://xcat.org/files/xcat/xcat-dep/2.x_Linux/xcat-dep-201702211334.tar.bz2

commit num: 25bbef0

TotalCase 244 Pass 217 Failed 27 NoRun 0

Failed cases: reg_linux_diskfull_installation_flat, go_xcat_local_repo_case7, go_xcat_noinput, go_xcat_with_x, go_xcat_with_xcat-version-1, go_xcat_online_repo_case6, xdsh_regular_command, xdsh_Q_command, xdsh_c_cn, xdsh_e_filename, xdsh_E, xdsh_t, xdsh_T, xdsh_o, psh_node_cmd_linux, psh_l, psh_i, prsync_dir_node, prsync_file_node, reg_linux_diskless_installation_flat, packimage_m_cpio_c_gzip, packimage_m_cpio_c_xz, packimage_m_tar_c_gzip, packimage_m_tar_c_xz, reg_linux_statelite_installation_flat, sles_migration1, sles_migration2,

````

````

======================================
  Test Result for Project rhels7.3-ppc64le-makedhcp-12
======================================

[rhels7.3+ppc64le]

Start time: 2017-05-10 06:01:30

End time  : 2017-05-10 11:19:05

All time consumption: 5 hours 17 minutes 35 seconds

Deploy test environment time consumption: 0 hours 10 minutes 51 seconds

xcat core : 2.13.3-snap201705090411

xcat dep  : http://xcat.org/files/xcat/xcat-dep/2.x_Linux/xcat-dep-201702211334.tar.bz2

commit num: 25bbef0

TotalCase 311 Pass 310 Failed 1 NoRun 0

Failed cases: confignics_config_multiple_port_withnicaliases_multiple_value,

````

````
======================================
  Test Result for Project ubuntu16.04.1-x86_64-makedhcp-2
======================================

[ubuntu16.04.1+x86_64]

Start time: 2017-05-12 04:02:22

End time  : 2017-05-12 12:55:12

All time consumption: 8 hours 52 minutes 50 seconds

Deploy test environment time consumption: 0 hours 8 minutes 40 seconds

xcat core : 2.13.3-snap201705090412

xcat dep  : http://xcat.org/files/xcat/xcat-dep/2.x_Ubuntu/xcat-dep-ubuntu-201702211344.tar.bz2

commit num: 25bbef0

TotalCase 278 Pass 250 Failed 28 NoRun 0

Failed cases: reg_linux_diskless_installation_flat, reg_linux_diskfull_installation_flat, go_xcat_local_repo_case7, go_xcat_noinput, go_xcat_with_xcat-version-1, go_xcat_online_repo_case6, packimage_m_cpio_c_gzip, packimage_m_cpio_c_pigz, packimage_m_cpio_c_xz, packimage_m_tar_c_pigz, packimage_m_tar_c_gzip, packimage_m_tar_c_xz, nodestat_noderange, xdsh_regular_command, xdsh_Q_command, xdsh_c_cn, xdsh_e_filename, xdsh_E, xdsh_t, xdsh_T, xdsh_o, psh_node_cmd_linux, psh_l, psh_i, prsync_dir_node, prsync_file_node, ubuntu_migration1_vm, ubuntu_migration2_vm,

````

Since the 2.13.perfopt diverged from master branch several weeks ago, so some of the failed cases are caused by the issues in master branch which were not fixed in that time. I will merge the 2.13.perfopt and master in my local repo and rerun the test to evaluate whether it's safe enough to merge in this week.

**Comment by immarvin** [comment]
The latest test result
````
======================================
  Test Result for Project sles12.2-ppc64le-makedhcp-7
======================================

[sles12.2+ppc64le]

Start time: 2017-05-15 07:58:25

End time  : 2017-05-15 15:40:27

All time consumption: 7 hours 42 minutes 2 seconds

Deploy test environment time consumption: 1 hours 41 minutes 52 seconds

xcat core : 2.13.4-snap201705150216

xcat dep  : http://xcat.org/files/xcat/xcat-dep/2.x_Linux/xcat-dep-201702211334.tar.bz2

commit num: 675b0ff

TotalCase 235 Pass 228 Failed 7 NoRun 0

Failed cases: makedns_d_node, makedns_n, makedns, xdsh_t, sles_migration1, sles_migration2, reg_linux_statelite_installation_flat, 


======================================
  Test Result for Project sles11.4-x86_64-makedhcp-4
======================================

[sles11.4+x86_64]

Start time: 2017-05-15 08:54:55

End time  : 2017-05-15 09:06:22

All time consumption: 0 hours 11 minutes 27 seconds

Deploy test environment time consumption: 0 hours 11 minutes 8 seconds

xcat core : 2.13.4-snap201705150216

xcat dep  : http://xcat.org/files/xcat/xcat-dep/2.x_Linux/xcat-dep-201702211334.tar.bz2

commit num: 675b0ff

TotalCase 239 Pass 0 Failed 1 NoRun 238

Failed cases: install_xCAT_on_rhels_sles, 

Final environment stop at install_xCAT_on_rhels_sles

======================================
  Test Result for Project rhels6.9-x86_64-makedhcp-5
======================================

[rhels6.9+x86_64]

Start time: 2017-05-15 05:02:12

End time  : 2017-05-15 08:53:21

All time consumption: 3 hours 51 minutes 9 seconds

Deploy test environment time consumption: 0 hours 8 minutes 37 seconds

xcat core : 2.13.4-snap201705150216

xcat dep  : http://xcat.org/files/xcat/xcat-dep/2.x_Linux/xcat-dep-201702211334.tar.bz2

commit num: 675b0ff

TotalCase 227 Pass 227 Failed 0 NoRun 0

Failed cases: 
======================================
  Test Result for Project rhels7.3-ppc64le-makedhcp-14
======================================

[rhels7.3+ppc64le]

Start time: 2017-05-15 02:43:16

End time  : 2017-05-15 07:57:02

All time consumption: 5 hours 13 minutes 46 seconds

Deploy test environment time consumption: 0 hours 11 minutes 36 seconds

xcat core : 2.13.4-snap201705150216

xcat dep  : http://xcat.org/files/xcat/xcat-dep/2.x_Linux/xcat-dep-201702211334.tar.bz2

commit num: 675b0ff

TotalCase 307 Pass 301 Failed 5 NoRun 1

Failed cases: confignics_config_multiple_port_withnicaliases_multiple_value, bmcdiscover_check_paswd, bmcdiscover_get_ipsource, redhat_migration1, redhat_migration2, 

Final environment stop at redhat_migration2

````



since the master branch is not locked anymore, I suggest to merge this branch to master branch and invoke the autotest of master branch on the DEV autotest environment.

File: perl-xCAT/xCAT/Utils.pm [hunk 1]
```diff
@@ -3314,6 +3316,19 @@ sub noderangecontainsMn
     # name
     my @mnames;   # management node names in the database, members of __mgmtnode
     my $tab = xCAT::Table->new('nodelist');
+    
+    my @nodelist = $tab->getAllAttribsWhere("node in ("."\'".join("\',\'", @noderange)."\'".") and groups like '%__mgmtnode%'",'node');
```
Comment (daniceexi) [inline perl-xCAT/xCAT/Utils.pm]: Intersting change, but we need be aware that this kind of change that using SQL directly might break the consistent support for multiple DB.

Now we don't need take care of DB2, but MariaDB, Postgresql and sqlite must be tested.

File: perl-xCAT/xCAT/NetworkUtils.pm [hunk 2]
```diff
@@ -270,69 +272,117 @@ sub getipaddr
...
+    my $isip=0;
+    if ($iporhost and ($iporhost =~ /\d+\.\d+\.\d+\.\d+/) || ($iporhost =~ /:/)){
+        $isip=1;
+    }
+
+
+#print "============================\n";
+#print Dumper(\%::hostiphash);
+#print "\n";
+#print Dumper(\%extraarguments);
```
Comment (daniceexi) [inline perl-xCAT/xCAT/NetworkUtils.pm]: Please refer to the ``man xcatdebug`` that we can use an easy way to enable and disable debug message:

```
Trace section
                ## TRACE_BEGIN
                # print "In the debug\n";
                ## TRACE_END

            Trace in a single line
                ## TRACE_LINE print "In the trace line\n";

```

File: perl-xCAT/xCAT/Utils.pm (hunk 1 above)
Comment (immarvin) [inline perl-xCAT/xCAT/Utils.pm]: hi @daniceexi ,the getAllAttribsWhere partly take care of the compatiblility. In this PR, I passed  the sql condition statement to getAllAttribsWhere, I  have verified the composed SQL command generated inside getAllAttribsWhere on mysql, sqlite and postgresql during the UT process

File: perl-xCAT/xCAT/NetworkUtils.pm (hunk 2 above)
Comment (immarvin) [inline perl-xCAT/xCAT/NetworkUtils.pm]: thanks for reminder @daniceexi  , I will utilize xcatdebug trace in the following refinement process

**Comment by whowutwut** [comment]
Should we delete this branch 2.13.perfopt if no longer needed?

**Comment by immarvin** [comment]
hi @whowutwut , it's ok to delete this branch on xcat2/xcat-core now

**Pull request merged**